app/src/main/java/com/ayush/weatherapp/home/HomePresenterImpl.java

```diff
@@ -168,27 +164,38 @@ private void stopLocationUpdates() {
   }
 
   private void fetchAddress(String latLng) {
+    // TODO refactor after mvp complete
+    Disposable disposable = geocodingRepository.getLocationDetails(latLng)
+        .subscribeOn(Schedulers.io())
+        .observeOn(AndroidSchedulers.mainThread())
+        .subscribeWith(getGeoLocationObserver());
+  }
 
-    Call<GeoLocation> geoLocationCall = geocodingAPIInterface.getLocationDetails(latLng);
-
-    geoLocationCall.enqueue(new Callback<GeoLocation>() {
-      @Override public void onResponse(@NonNull Call<GeoLocation> call,
-          @NonNull Response<GeoLocation> response) {
-        GeoLocation geoLocation = response.body();
-        List<Address> addressList = geoLocation.getAddresses();
+  @NonNull private DisposableObserver<GeoLocation> getGeoLocationObserver() {
+    return new DisposableObserver<GeoLocation>() {
+      @Override public void onNext(GeoLocation geoLocation) {
+        setLocation(geoLocation);
+      }
 
-        if (addressList != null && !addressList.isEmpty()) {
-          List<AddressComponents> addressComponentsList = addressList.get(0).getAddressComponents();
-          getView().setAddress(getAddress(addressComponentsList));
-        } else {
-          getView().setAddress(getString(R.string.not_available));
-        }
+      @Override public void onError(Throwable e) {
+        Timber.e(e);
       }
 
-      @Override public void onFailure(@NonNull Call<GeoLocation> call, @NonNull Throwable t) {
-        Timber.e(t);
+      @Override public void onComplete() {
       }
-    });
+    };
+  }
+
+  private void setLocation(GeoLocation geoLocation) {
+    List<Address> addressList = geoLocation.getAddresses();
+
+    if (addressList != null && !addressList.isEmpty()) {
+      List<AddressComponents> addressComponentsList =
+          addressList.get(0).getAddressComponents();
+      getView().setAddress(getAddress(addressComponentsList));
```

Comment:
getAddress function can check for null and empty and return the string from else block.

Reply:
:ok: